nexus/src/internal_api/http_entrypoints.rs

```diff
@@ -280,3 +285,76 @@ async fn cpapi_metrics_collect(
         .instrument_dropshot_handler(&request_context, handler)
         .await
 }
+
+#[derive(Deserialize, JsonSchema)]
+struct AllPath {
+    path: String,
+}
+
+/// Endpoint used by Sled Agents to download cached artifacts.
+#[endpoint {
+    method = GET,
+    path = "/artifacts/{path}",
+}]
+async fn cpapi_artifact_download(
+    request_context: Arc<RequestContext<Arc<ServerContext>>>,
+    path: Path<AllPath>,
+) -> Result<Response<Body>, HttpError> {
+    let context = request_context.context();
+    let nexus = &context.nexus;
+    let mut entry = PathBuf::from(BASE_ARTIFACT_DIR);
+
+    // TODO: Most of the below code is ready to accept a multi-component path,
+    // such as in:
+    // https://github.com/oxidecomputer/dropshot/blob/78be3deda556a9339ea09f3a9961fd91389f8757/dropshot/examples/file_server.rs#L86-L89
+    //
+    // However, openapi does *not* like that currently, so we limit the endpoint
+    // to only accepting single-component paths.
+    let path = vec![path.into_inner().path];
+
+    for component in &path {
+        // Dropshot should not provide "." and ".." components.
+        assert_ne!(component, ".");
+        assert_ne!(component, "..");
```

Comment:
does this panic? if so should we return an error instead

Reply:
I have a test for this - it should be filtered out by dropshot, but this is me being paranoid about that point.

I didn't return an error because "I don't think it should be possible to hit this case", hence the assert.